### PR TITLE
docker validation improvement

### PR DESCRIPTION
**Pull request by jordarlu**
Description
This PR is to address https://github.com/opensearch-project/opensearch-build/issues/3205

Issues Resolved

- Adding a flag/argument to do Only the digest validaiton without moving to run cluster, nor the API test
- Adding a flag/argument to use Only staging images ( in which, we will not check digest )
- Adding a max retry for testing if the OpenSearch cluster is ready for accepting HTTP request
- Return 0 instead of True to the caller when the script ends successfully

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by jordarlu** [comment]
Hi, the update on api_test will not impact others who use this method because the input and output format staying the same as before.
In src/run_validation.py, the test_result will be either True or False which taken from the actual job runner, so, I transform it to 0 for True, and 1 for False.
This PR also has two additional arguments added to commit with feedback from dry-run test a while ago, pls see the output for those two additional optional argument below.
The inplace_change makes the changes on copied docker-compose.yml at temp directory. The source file is not touched.

**Comment by jordarlu** [comment]
My apologies on making the review diffcult by replacing a new PR with the previous one ..if it helps, here is the link to the previosu PR https://github.com/opensearch-project/opensearch-build/pull/3264.  Thank you for speding extra effort on reviewing it ...

File: src/validation_workflow/docker/validation_docker.py
Comment (gaiksaya) [inline src/validation_workflow/docker/validation_docker.py:50]: I did not get the True/False mapping. Can you explain? AFAI understand, its whether its staging or prod. Would ask to name the key of the mapping accordingly. True/False is confusing or add comments.
Adding @prudhvigodithi / @peterzhuamazon if they can suggest something better.

File: src/validation_workflow/docker/validation_docker.py
Comment (gaiksaya) [inline src/validation_workflow/docker/validation_docker.py:78]: Log something here instead of `pass` like what process was completed.

File: src/validation_workflow/docker/validation_docker.py
Comment (gaiksaya) [inline src/validation_workflow/docker/validation_docker.py:89]: nit: typo in downloaed. Should be downloaded/local

File: src/validation_workflow/docker/validation_docker.py
Comment (gaiksaya) [inline src/validation_workflow/docker/validation_docker.py:124]: Do we need to warn and raise exception both?

File: src/validation_workflow/docker/validation_docker.py
Comment (gaiksaya) [inline src/validation_workflow/docker/validation_docker.py:143]: Wondering if we want the `curl` to be silent "-s". Let it be loud. Even if it fails exact error can be viewed then.

File: src/validation_workflow/docker/validation_docker.py
Comment (gaiksaya) [inline src/validation_workflow/docker/validation_docker.py:159]:
```suggestion
                logging.info('cleanup is not complete')
```

File: src/validation_workflow/docker/validation_docker.py
Comment (gaiksaya) [inline src/validation_workflow/docker/validation_docker.py:156]:
```suggestion
                logging.info('cleanup is complete')
```

File: src/validation_workflow/docker/validation_docker.py
Comment (gaiksaya) [inline src/validation_workflow/docker/validation_docker.py:165]:
```suggestion
            logging.info('Docker validation cleanup is complete')
```

File: src/validation_workflow/docker/validation_docker.py
Comment (gaiksaya) [inline src/validation_workflow/docker/validation_docker.py:176]:
```suggestion
            logging.error("Error: %s - %s." % (e.filename, e.strerror))
```
Also raise an exception that just logging.

File: src/validation_workflow/docker/validation_docker.py
Comment (gaiksaya) [inline src/validation_workflow/docker/validation_docker.py:231]: Raise an exception than just printing it.

File: src/validation_workflow/docker/validation_docker.py
Comment (gaiksaya) [inline src/validation_workflow/docker/validation_docker.py:233]: Log what process completed rather than pass

**Review by gaiksaya — CHANGES_REQUESTED** [review]
Since the validation depends heavily on docker-compose, please add tests ensuring that those URLs are reachable. If someone raises a PR to delete those files, docker validation test cases should fail.

File: src/validation_workflow/docker/validation_docker.py
Comment (jordarlu) [reply src/validation_workflow/docker/validation_docker.py:50]: thanks, @gaiksaya.. because we only have Staging and Production, and I want to determine which one to apply by one of the input argument, 'self.args.using_staging_artifact_only', which will be either True or False; hence, I wanted to map Staging = True, and Production = False. 
Therefore, at L56~57, I can apply to the correct URL based on the input argument 'self.args.using_staging_artifact_only'.

File: src/validation_workflow/docker/validation_docker.py
Comment (jordarlu) [reply src/validation_workflow/docker/validation_docker.py:78]: got it ..it makes sense ..let me update it in next commit

File: src/validation_workflow/docker/validation_docker.py
Comment (jordarlu) [reply src/validation_workflow/docker/validation_docker.py:89]: oops ./ thanks for the correction .. let me get that right in my next commit

File: src/validation_workflow/docker/validation_docker.py
Comment (jordarlu) [reply src/validation_workflow/docker/validation_docker.py:143]: sure, @gaiksaya .. I wanted to slience it because there was too much noises, informations and caused the log/output difficult to read, but I can certainly remove the "-s" ...

File: src/validation_workflow/docker/validation_docker.py
Comment (jordarlu) [reply src/validation_workflow/docker/validation_docker.py:156]: :) got it!!

File: src/validation_workflow/docker/validation_docker.py
Comment (jordarlu) [reply src/validation_workflow/docker/validation_docker.py:176]: got it !! makes sense.

File: src/validation_workflow/docker/validation_docker.py
Comment (jordarlu) [reply src/validation_workflow/docker/validation_docker.py:124]: mm... I wanted to also show a simple string on the output as well ....but yes,, I can certainly remove it ... let me do that .!! thanks!!

**Comment by jordarlu** [comment]
> Since the validation depends heavily on docker-compose, please add tests ensuring that those URLs are reachable. If someone raises a PR to delete those files, docker validation test cases should fail.

HI, @gaiksaya ..just want to make sure I got it right, you meant to have a test as validation on file existing of those two docker-compose-1/2.x.yml in build repo.... correct? thanks

**Comment by gaiksaya** [comment]
> > Since the validation depends heavily on docker-compose, please add tests ensuring that those URLs are reachable. If someone raises a PR to delete those files, docker validation test cases should fail.
> 
> HI, @gaiksaya ..just want to make sure I got it right, you meant to have a test as validation on file existing of those two docker-compose-1/2.x.yml in build repo.... correct? thanks

Yes

File: src/validation_workflow/docker/validation_docker.py
Comment (jordarlu) [reply src/validation_workflow/docker/validation_docker.py:50]: Myabe we can change to something like this?

**Comment by jordarlu** [comment]
Hi, @gaiksaya , if you can have another round of review on updated portion, thanks !!

File: src/validation_workflow/api_request.py
Comment (dblock) [inline src/validation_workflow/api_request.py:24]: Why are we replacing a native HTTP implementation with `curl`?

File: src/validation_workflow/docker/validation_docker.py
Comment (dblock) [inline src/validation_workflow/docker/validation_docker.py:204]: Use inheritance. This should be refactored into different validation implementations, one for staging and another for otherwise.

File: src/validation_workflow/docker/validation_docker.py
Comment (dblock) [inline src/validation_workflow/docker/validation_docker.py:230]: There's no need to catch and re-raise, just let it raise an exception.

File: src/validation_workflow/docker/validation_docker.py
Comment (dblock) [inline src/validation_workflow/docker/validation_docker.py:32]: move to a constant

File: src/validation_workflow/api_request.py
Comment (jordarlu) [reply src/validation_workflow/api_request.py:24]: Thanks for the comment, @dblock .. a few weeks back, when I was doing testing with @peterzhuamazon , we found out that the Curl is more stable on privding the result than using the request lib for some reasons.

File: src/validation_workflow/docker/validation_docker.py
Comment (jordarlu) [reply src/validation_workflow/docker/validation_docker.py:230]: thanks, sure, let me update it in my next commit..

File: src/validation_workflow/docker/validation_docker.py
Comment (jordarlu) [reply src/validation_workflow/docker/validation_docker.py:38]: Thanks for the comment, @dblock , the new code hopefully explains more on selecting the image source between dockerhub and ecr, and plus supporting staging and production...
the code will choose either dockerhub or ecr ; and either using staging-only or comparing between staging and production before briing up a container, and it will be based on the arguemnts taken from the user input

sorry that I am lost when you comment on 'pass the image name dynamically' ... I guess you don't mean to have user to input the full image name as an argument if I am not mistaken .. thanks,

**Review by dblock — COMMENTED** [review]
I made some directional suggestions below. Break up the code into functions and use inheritance so that we don't have too many different things going on in the same function x2 with opensearch and opensearch dashboards

File: src/validation_workflow/api_request.py
Comment (jordarlu) [reply src/validation_workflow/api_request.py:24]: Thanks, dB, I will keep using the requests libarary. so, no change on this script.

File: src/validation_workflow/api_test_cases.py
Comment (jordarlu) [reply src/validation_workflow/api_test_cases.py:42]: It was originally putting a delay between each test, not required anymore . .let me remove it .. thanks for the comment

File: src/validation_workflow/docker/validation_docker.py
Comment (jordarlu) [reply src/validation_workflow/docker/validation_docker.py]: Thanks for the meeting, and comments, @dblock 
I made some updates.. would you mind have a look again? Thanks!!

File: src/validation_workflow/docker/validation_docker.py
Comment (jordarlu) [reply src/validation_workflow/docker/validation_docker.py:38]: Thanks, @dblock ..I made an update just now .. hopefully it will work betteer this time..

File: src/validation_workflow/docker/validation_docker.py
Comment (dblock) [inline src/validation_workflow/docker/validation_docker.py:186]: You should make this OO. Create a class called `DockerImage` that takes source, artifact and staging as input and generates the name as a method and returns `public.ecr.aws/opensearchstaging/opensearch-dashboards` and such as a property. Each one of the parts of the result has meaning. For example a method called `host` would return `null` if the registry is dockerhub, public.ecr.aws if it's ecr, etc.

File: src/validation_workflow/docker/validation_docker.py
Comment (dblock) [inline src/validation_workflow/docker/validation_docker.py:43]: This code should be

```python
self.images = DockerImages(["OpenSearch, "OpenSearch Dashboards"])
self.images.download()
```

You're missing a collection class `DockerImages` and an individual class `DockerImage`.

**Review by dblock — APPROVED** [review]
This is still very _not_ object oriented with a lot of `if`'s and exception catching. 
I think you should fix the parameter `_` into `-` and let's merge this, then see if you can take some time to refactor it further after that?

**Comment by jordarlu** [comment]
> This is still very _not_ object oriented with a lot of `if`'s and exception catching. I think you should fix the parameter `_` into `-` and let's merge this, then see if you can take some time to refactor it further after that?

Thanks for the approval, @dblock .. yes .. I will definitely optimit it later on with using aoother PR .. and thanks for the chatting as well

**Comment by jordarlu** [comment]
> Since the validation depends heavily on docker-compose, please add tests ensuring that those URLs are reachable. If someone raises a PR to delete those files, docker validation test cases should fail.

Thank you, @gaiksaya , if you can give another review..

**Review by gaiksaya — APPROVED** [review]
Hope to see a follow up PR with changes suggested by dB!

File: src/validation_workflow/validation_args.py
Comment (gaiksaya) [reply src/validation_workflow/validation_args.py:89]: @jordarlu Do you want to think small thing in before merge?

**Comment by jordarlu** [comment]
> Hope to see a follow up PR with changes suggested by dB!

Certainly , let me create an issue for following up on the comments from dB  :)    thanks again!!

File: src/validation_workflow/validation_args.py
Comment (jordarlu) [reply src/validation_workflow/validation_args.py:89]: sorry,, I missed it ... thanks for notifying me this .. and yes... let me update it really quick before merging ...and will create another issue to follow up on comments for further improvements .. thanks !!

**Comment by jordarlu** [comment]
https://github.com/opensearch-project/opensearch-build/issues/3475 is opened to track the enhancement

**Pull request merged**
